### PR TITLE
fix(developer): ldml: fix for unescaping Unicode values 🙀

### LINKED ISSUE #7569: bug(developer): ldml does not support escaping 🙀
```xml
    <display to="\u{0300}" display="`" />
```

should go to U+0300  etc.

Note:
- [ ] see and resolve https://github.com/keymanapp/keyman/pull/7570#issuecomment-1372602027 before closing this once #7630 is merged. 

#7054

### PR DESCRIPTION
**Pull request by srl295**
fix(developer): ldml: add escaping 🙀

- common util unescapeString with its own tests
- add a test case using the disp element

Fixes #7569

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/kmc-keyboard/test/fixtures/basic.xml
Comment (srl295) [inline developer/src/kmc-keyboard/test/fixtures/basic.xml:21]: Hm, maybe don't need this in basic also.

**Comment by srl295** [comment]
test case is

```xml
    <display to="\u{0300}" display="`" />
    <displayOptions baseCharacter="x"/>
  </displays>
```

but shouldn't be in basic.xml.. yet

File: common/web/types/src/util/util.ts
Comment (mcdurdin) [inline common/web/types/src/util/util.ts:19]:
```suggestion
const MATCH_HEX_ESCAPE = /\\u{([0-9a-fA-F]{1,6})}/g;
```

`\u{10FFFF}` has 6 digits.

Query: `\u{a}` doesn't seem like it should be acceptable? Do we want to allow control characters (below 0x20) in these strings? This perhaps should be handled outside this function though.

File: common/web/types/src/util/util.ts
Comment (mcdurdin) [inline common/web/types/src/util/util.ts:26]: What does this comment mean?

File: common/web/types/test/util/test-unescape.ts
Comment (mcdurdin) [inline common/web/types/test/util/test-unescape.ts:26]:
```suggestion
    assert.equal(unescapeString('\\u{1e109}'), '\u{1e109}');
    assert.equal(unescapeString('\\u{10fff0}'), '\u{10fff0}');
```

Suggest using the ES6 escaped strings on the expecteds for clarity.

File: common/web/types/test/util/test-unescape.ts
Comment (mcdurdin) [inline common/web/types/test/util/test-unescape.ts:21]:
```suggestion
  it("should correctly handle 1..6 char escapes", function() {
```

File: common/web/types/src/util/util.ts
Comment (mcdurdin) [inline common/web/types/src/util/util.ts:24]: This will throw `RangeError` on invalid values such as `\u{110000}`. How do we want to handle that?

File: common/web/types/src/util/util.ts
Comment (mcdurdin) [inline common/web/types/src/util/util.ts:22]:
```suggestion
  if(!s) {
    return s;
  }
```

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Looking good; a couple of changes requested.

File: common/web/types/src/util/util.ts
Comment (srl295) [reply common/web/types/src/util/util.ts:26]:
```suggestion
  return s;
```

it was a leftover!

File: common/web/types/src/util/util.ts
Comment (srl295) [reply common/web/types/src/util/util.ts:24]: Perhaps it should be caught here and thrown as its own `CompilerError` specifically stating an invalid escape.

Comment (mcdurdin) [reply common/web/types/src/util/util.ts:24]: That would be good. Then the compiler will handle `CompilerError`s and reframe them through the compiler logging. A real `RangeError` elsewhere will then be unhandled, as it should be.

We also want to block surrogate pairs and illegal chars somewhere, but not here.

File: common/web/types/src/kmx/kmx-plus.ts
Comment (srl295) [reply common/web/types/src/kmx/kmx-plus.ts:136]: That is a good question. I will double check.

Actually I can think of one class of cases:  Where the NMTOKEN(S) are being included, such as where the literal key ID is pulled in.

**Comment by srl295** [comment]
@mcdurdin I think I have addressed the main code issues, except for:

- [ ] verify what should be unescaped and what shouldn't.

When I do that I'll move this out of draft.  The regex I wanted to write didn't work, yet.  So I am using a simpler one.

**Comment by srl295** [comment]
One issue I see is that, for example, `longPress` needs to be escaped, but this PR is at the bottom of the pile right now and the `layr` PR is closer to the top.  So this may take a couple iterations until escaping is in the right place.  I will see if I can add a DTD annotation of some sort to the spec to make it easier to sort these out.

**Comment by srl295** [comment]
Over on the DTD side, I'm proposing a new annotation for escaping: https://github.com/unicode-org/cldr/pull/2642 It's already in the text of the spec, but this makes it easier to see at a glance which attributes allow unescaping.

**Comment by srl295** [comment]
this is now based on #7630  and so I can bring in the rest of the escaped attributes

**Comment by srl295** [comment]
OK.  Pulled in the last  few attributes that needed escaping.  With unit tests.

File: common/web/types/src/kmx/kmx-plus.ts
Comment (srl295) [reply common/web/types/src/kmx/kmx-plus.ts]: I think this is now addressed.

**Comment by mcdurdin** [comment]
I might wait until you have resolved merge conflicts before reviewing?

File: common/web/types/src/util/util.ts
Comment (srl295) [reply common/web/types/src/util/util.ts]: it's thrown as an `UnescapeError` now.

**Comment by srl295** [comment]
@mcdurdin as i noted i'm throwing an `EscapingError` if escaping goes awry.  Should that be caught in `Compiler.compile()` perhaps? Since it's cross module I didn't see a good way to get at `callbacks.reportMessage()` from util down in common/web/types.  Maybe a future PR, leaving this issue open?

File: developer/src/kmc-keyboard/test/test-keys.ts
Comment (mcdurdin) [inline developer/src/kmc-keyboard/test/test-keys.ts:25]: Perhaps should use escaped value here as it has notdef glyph for me.

**Review by mcdurdin — APPROVED** [review]
LGTM tiny nit only

**Pull request merged**
